scanner/client/client.go

```diff
@@ -0,0 +1,189 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/cenkalti/backoff/v3"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/quay/zlog"
+	v4 "github.com/stackrox/rox/generated/internalapi/scanner/v4"
+	"github.com/stackrox/rox/pkg/clientconn"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ScannerClient is the interface that contains the StackRox Scanner
+// application-oriented methods. It's offered to simplify application code to
+// call StackRox Scanner.
+type ScannerClient interface {
+	// GetOrCreateImageIndex first attempts to get an existing index report for the
+	// image reference, and if not found or invalid, it then attempts to index the
+	// image and return the generated index report if successful, or error.
+	GetOrCreateImageIndex(ctx context.Context, ref name.Digest, auth authn.Authenticator) (*v4.IndexReport, error)
+
+	// IndexAndScanImage scans an image for vulnerabilities. If the index report
+	// for that image does not exist, it is created. It returns the vulnerability
+	// report.
+	IndexAndScanImage(context.Context, name.Digest, authn.Authenticator) (*v4.VulnerabilityReport, error)
+
+	// Close cleans up any resources used by the implementation.
+	Close() error
+}
+
+// grpcClient A scanner client implementation based on gRPC endpoints.
+type grpcClient struct {
+	indexer        v4.IndexerClient
+	matcher        v4.MatcherClient
+	grpcConnection *grpc.ClientConn
+}
+
+// NewGRPCScannerClient creates a new gRPC scanner client.
+func NewGRPCScannerClient(ctx context.Context, opts ...Option) (ScannerClient, error) {
+	o := makeOptions(opts...)
+	connOpt := clientconn.Options{
+		TLS: clientconn.TLSConfigOptions{
+			GRPCOnly:           true,
+			InsecureSkipVerify: true,
+		},
+	}
+	if o.withTLSVerify {
+		ca, err := o.GetCA()
+		if err != nil {
+			return nil, fmt.Errorf("creating CA: %w", err)
+		}
+		connOpt.TLS.InsecureSkipVerify = false
+		connOpt.TLS.RootCAs = ca.CertPool()
+		connOpt.TLS.UseClientCert = clientconn.MustUseClientCert
+		connOpt.TLS.ServerName = o.serverName
+	}
+	conn, err := clientconn.GRPCConnection(ctx, o.mtlsSubject, o.address, connOpt)
+	if err != nil {
+		return nil, err
+	}
+	indexerClient := v4.NewIndexerClient(conn)
+	matcherClient := v4.NewMatcherClient(conn)
```

Comment:
How do we know the Scanner we want to talk to is running both of these servers? I guess these two functions never return `nil`, but I just want to be sure we don't `panic` if the server is not running in `matcher` mode for example